### ISSUE TITLE
Extract bifurcation angles from an L5TTPC1 population

### ISSUE THREAD

**Issue by chalimou** [BlueBrain/NeuroM#194]
Hi,

I created 100 L5TTPC1 morphologies with morphsyn in L5TTPC1_nancy. When I run the script extract_feature.py (see below), I get the following problem. The interesting bit is that the problem exists only for TreeType.all and NOT for TreeType.basal_dendrite. For the first morphology, isolated in the directory L5TTPC1_one everything works fine, too!

nancy

```python
In [26]: data_dir
Out[26]: '/Users/chalimou/Projects/morphsyn/build/L5TTPC1_nancy/'

In [27]: feature
Out[27]: 'local_bifurcation_angles'

In [28]: res = extract_feature(data_dir,feature,TreeType.all)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-28-a98154e61e39> in <module>()
----> 1 res = extract_feature(data_dir,feature,TreeType.all)

/Users/chalimou/Projects/NEUROM/NeuroM/apps/extract_feature.py in extract_feature(data_dir, feature, neurite_type)
      6     population = load_neurons(data_dir)
      7
----> 8     feature_data = [getattr(n, 'get_' + feature)(neurite_type=neurite_type) for n in population]
      9
     10     return feature_data

/Users/chalimou/Projects/NEUROM/NeuroM/neurom/ezy/neuron.pyc in get_local_bifurcation_angles(self, neurite_type)
    150         '''
    151         return self.neurite_loop(i_local_bifurcation_angle,
--> 152                                  neurite_type=neurite_type)
    153
    154     def get_remote_bifurcation_angles(self, neurite_type=TreeType.all):

/Users/chalimou/Projects/NEUROM/NeuroM/neurom/ezy/neuron.pyc in neurite_loop(self, iterator_type, mapping, neurite_type)
    303             Iterable containing the iteration targets after mapping.
    304         '''
--> 305         return self._pkg(self.iter_neurites(iterator_type, mapping, neurite_type))
    306
    307     def _pkg(self, iterator):

/Users/chalimou/Projects/NEUROM/NeuroM/neurom/ezy/neuron.pyc in _pkg(self, iterator)
    307     def _pkg(self, iterator):
    308         '''Create an iterable from an iterator'''
--> 309         return self._iterable_type([i for i in iterator])
    310
    311     def _compare_neurites(self, other, neurite_type, comp_function=compare_trees):

/Users/chalimou/Projects/NEUROM/NeuroM/neurom/analysis/morphtree.pyc in local_bifurcation_angle(bifurcation_point)
     55     return mm.angle_3points(bifurcation_point.value,
     56                             bifurcation_point.children[0].value,
---> 57                             bifurcation_point.children[1].value)
     58
     59

/Users/chalimou/Projects/NEUROM/NeuroM/neurom/analysis/morphmath.pyc in angle_3points(p0, p1, p2)
     89     vec2 = vector(p2, p0)
     90     return math.acos(np.dot(vec1, vec2) /
---> 91                      (np.linalg.norm(vec1) * np.linalg.norm(vec2)))
     92
     93

ValueError: math domain error

In [29]: res = extract_feature(data_dir,feature,TreeType.basal_dendrite)

In [30]: data_dir = '/Users/chalimou/Projects/morphsyn/build/L5TTPC1_one/'

In [31]: res = extract_feature(data_dir,feature,TreeType.all)

In [32]: res = extract_feature(data_dir,feature,TreeType.basal_dendrite)
```

-------------------------
extract_feature.py:
```python
from neurom.ezy import load_neurons
from neurom.core.types import TreeType

def extract_feature(data_dir, feature, neurite_type=TreeType.all):

    population = load_neurons(data_dir)

    feature_data = [getattr(n, 'get_' + feature)(neurite_type=neurite_type) for n in population]

    return feature_data


def flatten(feature_data):
    from itertools import chain
    return list(chain(*feature_data))
```

**Comment by juanchopanza** [comment]
Did you manage to reproduce the problem with a single file? If so, could you send it? If not, could you put the data somewhere I can access it (for example, nfs?)

**Comment by chalimou** [comment]
Hi, this is exactly the problem :) I just put the first cell on a separate directory and run it but everything works fine for that! It's only for the whole population of 100 cells that TreeType.all does not work. TreeType.basal_dendrite works for the population. I'll now try small populations and let you know. Nancy

**Comment by chalimou** [comment]
Now, I think I localised the cell that causes the problems: It's the 37th one. if you run the script:

--------------------------
from neurom.ezy import load_neurons
from neurom.core.types import TreeType

def extract_feature(data_dir, feature, neurite_type=TreeType.all):
    population = load_neurons(data_dir)
    feature_data = [getattr(n, 'get_' + feature)(neurite_type=neurite_type) for n in population]
    return feature_data
-----------------------------
with this single cell, feature = 'local_bifurcation_angles' it gives the math domain error.
[L5TTPC1_37th.zip](https://github.com/BlueBrain/NeuroM/files/78216/L5TTPC1_37th.zip)

I hoped it uploaded the file correctly, otherwise I can send it per mail. I try to learn to use those systems :)

**Comment by juanchopanza** [comment]
Sorry, I cannot reproduce the problem. Here is a simple script that attempts to reproduce what I think you did. For the record, this is the kind of code you should submit. A script that can be simply run to reproduce the reported issue.

```python
from neurom.ezy import load_neurons
from neurom.core.types import TreeType

def extract_feature(data_dir, feature, neurite_type=TreeType.all):
    population = load_neurons(data_dir)
    feature_data = [getattr(n, 'get_' + feature)(neurite_type=neurite_type)
                    for n in population]
    return feature_data

if __name__ == '__main__':
    # data is ./L5TTPC1_37th/L5TTPC1_nancy_ID36.h5
    # from L5TTPC1_37th.zip
    ddir = 'L5TTPC1_37th'
    feature = 'local_bifurcation_angles'
    f = extract_feature(ddir, feature, TreeType.all)
    print f
```

**Comment by chalimou** [comment]
Hi Juan, can you do me a favour? I think I sent you the wrong cell. It's the cell with ID35 that causes me the problems and not with ID36. I'm attaching the script test.py and the ID35 cell below. Could you run it in your environment  when you find a minute? Thanks very much and sorry for the mess :) nancy
[cell35_script.zip](https://github.com/BlueBrain/NeuroM/files/78719/cell35_script.zip)

**Comment by juanchopanza** [comment]
I can reproduce it now. I will have a look at it.

**Comment by juanchopanza** [comment]
I found the problem: `morphmath.angle_3points` is not robust against collinear vectors. I changes the implementation for a more robust one (after adding obvious tests that should have been there in the first place). See pull request #195.

**Comment by chalimou** [comment]
thanks very much Juan! I think there are no embarrassing bugs, embarrassing was to send you the wrong cell and let you search in the wrong place :)

nancy
On 06 Jan 2016, at 10:04, Juan Palacios <notifications@github.com<mailto:notifications@github.com>> wrote:


I found the problem: morphmath.angle_3points is not robust against collinear vectors. I changes the implementation for a more robust one (after adding obvious tests that should have been there in the first place). See pull request #195<https://github.com/BlueBrain/NeuroM/pull/195>.

**Comment by juanchopanza** [comment]
Fixed with #195.
